homeassistant/components/rainmachine/button.py

```diff
@@ -0,0 +1,92 @@
+"""Buttons for the RainMachine integration."""
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from dataclasses import dataclass
+
+from regenmaschine.controller import Controller
+from regenmaschine.errors import RainMachineError
+
+from homeassistant.components.button import (
+    ButtonDeviceClass,
+    ButtonEntity,
+    ButtonEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.entity import EntityCategory
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import RainMachineData, RainMachineEntity
+from .const import DATA_PROVISION_SETTINGS, DOMAIN
+from .model import RainMachineEntityDescription
+
+
+@dataclass
+class RainMachineButtonDescriptionMixin:
+    """Define an entity description mixin for RainMachine buttons."""
+
+    push_action: Callable[[Controller], Awaitable]
+
+
+@dataclass
+class RainMachineButtonDescription(
+    ButtonEntityDescription,
+    RainMachineEntityDescription,
+    RainMachineButtonDescriptionMixin,
+):
+    """Describe a RainMachine button description."""
+
+
+BUTTON_KIND_REBOOT = "reboot"
+
+
+async def _async_reboot(controller: Controller) -> None:
+    """Reboot the RainMachine."""
+    await controller.machine.reboot()
+
+
+BUTTON_DESCRIPTIONS = (
+    RainMachineButtonDescription(
+        key=BUTTON_KIND_REBOOT,
+        name="Reboot",
+        api_category=DATA_PROVISION_SETTINGS,
+        push_action=_async_reboot,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up RainMachine buttons based on a config entry."""
+    data: RainMachineData = hass.data[DOMAIN][entry.entry_id]
+
+    async_add_entities(
+        [
+            RainMachineButton(entry, data, description)
+            for description in BUTTON_DESCRIPTIONS
+        ]
+    )
+
+
+class RainMachineButton(RainMachineEntity, ButtonEntity):
+    """Define a RainMachine button."""
+
+    _attr_device_class = ButtonDeviceClass.RESTART
+    _attr_entity_category = EntityCategory.CONFIG
+
+    entity_description: RainMachineButtonDescription
+
+    async def async_press(self) -> None:
+        """Send out a restart command."""
+        try:
+            await self.entity_description.push_action(self._data.controller)
+        except RainMachineError as err:
+            raise HomeAssistantError(
+                f'Error while pressing button "{self.entity_id}": {err}'
+            ) from err
+
+        async_dispatcher_send(self.hass, self.coordinator.signal_reboot_requested)
```

Comment:
Why does the integration need to know about the reboot?

Reply:
See comment [above](#issuecomment-1203021577). In short:

1. We have multiple coordinators, each of which needs to react to a reboot by marking its associated entities as `unavailable` (indirectly, of course).
2. One coordinator (the one with the shortest update interval) waits until it can ping the API again; once it can, it "notifies" the other coordinators. This ensures that all the entities come back online simultaneously (vs. in the piecemeal fashion that would occur with each coordinator having a different update interval).

This is precisely the same as what we merged for [Guardian](https://github.com/home-assistant/core/pull/75234) (except in that case, each coordinator has the same update interval, so we allow their entities to come back online naturally).

Reply:
Since they don't update at the same interval, its possible that the shortest one won't see the reboot first because it happens to align to update later than one of the other ones. ie. it gets unluckly.

Should all of them watch for reboot and whoever sees the reboot first sends the signal?

Reply:
I thought about that... Thoughts on whether I'd have to `asyncio.Lock` that somehow? Trying to envision a scenario where coordinator A comes back and notifies, but then right after, coordinator B does the same. Ultimately won't damage anything, but potentially lots of extra traffic.

Reply:
Does it matter which coordinator sends the signal first as long as you only accept the first one?

Reply:
No, doesn't matter. 👍🏻